### DISCUSSION TITLE
MOOSE Compile Problem

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#17127] [Q&A Getting Started] [answered]
Hello:  this is a first.  I have never had MOOSE fail to compile before.

System is a MacBook Pro running 10.14.6
Xcode version 10.2.1

First, I updated my moose environment via conda.

I cloned moose into ~/projects/moose

I only needed the NS app, so:  

(moose) joe$ cd /Users/joe/projects/moose/modules/navier_stokes 
(moose) joe$ make -j6

It ran for a while but then I get an error:

In file included from /Users/joe/projects/moose/framework/build/unity_src/mesh_Unity.C:9:
/Users/joe/projects/moose/framework/src/mesh/MooseMesh.C:2491:12: error: no member named 'set_isnt_prepared' in 'libMesh::MeshBase'
    _mesh->set_isnt_prepared();
    ~~~~~  ^
/Users/joe/projects/moose/framework/src/mesh/MooseMesh.C:2797:28: error: too many arguments to function call, expected at most single argument 'os', have 2 arguments
  getMesh().print_info(os, verbosity);
  ~~~~~~~~~~~~~~~~~~~~     ^~~~~~~~~
/Users/jmkelly/miniconda3/envs/moose/libmesh/include/libmesh/mesh_base.h:1193:3: note: 'print_info' declared here
  void print_info (std::ostream & os=libMesh::out) const;
  ^
2 errors generated.
make: *** [/Users/joe/projects/moose/framework/build/unity_src/mesh_Unity.x86_64-apple-darwin19.5.0.opt.lo] Error 1
make: *** Waiting for unfinished jobs....

Any suggestions?  The entire terminal output is attached.  Thanks,
  Joe Kelly
[compile_error.txt](https://github.com/idaholab/moose/files/6039126/compile_error.txt)

**Comment by cticenhour** [comment, marked as answer]
This is probably due to an old copy of libmesh, caused by a failure of conda to update past a certain version. 

What is the version number for `moose-libmesh` when you perform the command `conda list`? If that version number leads with `2020.12.xx` then we need to force the update to the newest version. In your moose environment, run the command `conda install moose-libmesh=2021.02.15` which will force an update to the newest version of libmesh. Then perform a clean (`make clobberall`) and try to `make` again. 

## TL;DR on the thread below:

This ended up being three main things...
(1) An out-of-date `moose-libmesh` package causing the original compile issue
(2) An out-of-date conda channel (`https://mooseframework.org/conda/moose`) which needed to be updated to `idaholab`
(3) `root` user ownership of `~/.conda`, which needs to be owned by the regular user in order to manipulate environments. 

As a general reminder to all users - do not use `sudo` while making changes or updates to conda if it is installed in your home directory! It can cause permissions issues that can lead to a broken or partially-broken conda installation.

**Reply by joe61vette** [reply]
That does appear to be a problem.  I got:

moose-libmesh             2020.10.05 

So, I'll go ahead and try to force it to update.  Thanks!

**Reply by sapitts** [reply]
FWIW, this procedure helped me solve a similar issue this morning

**Reply by joe61vette** [reply]
Unfortunately, the Conda command didn't work.  I got:

(base) JMKs-MBP:~ jmkelly$ conda activate moose
(moose) JMKs-MBP:~ jmkelly$ conda install moose-libmesh=2021.02.15
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: | failed

CondaError: KeyboardInterrupt

I say there forever, so I finally killed it.

**Reply by cticenhour** [reply]
OK - so in that case, I would create another environment if the database is confused or corrupted. You can do that by performing:
`conda remove -n moose --all` 
and then 
`conda create -n moose moose-libmesh moose-tools`
to get back to a basic moose environment.

**Reply by cticenhour** [reply]
Conda might complain if you try to remove the moose environment without deactivating first. So if it does, do `conda deactivate` to get back to the base environment.

**Reply by joe61vette** [reply]
Something must really be messed up.  I did:

conda deactivate
conda remove -n moose --all

it listed packages to be removed and I replied y.  Then:

Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: - WARNING conda.core.path_actions:verify(963): Unable to create environments file. Path not writable.
  environment location: /Users/jmkelly/.conda/environments.txt

done
Executing transaction: \ WARNING conda.core.envs_manager:register_env(52): Unable to register environment. Path not writable or missing.
  environment location: /Users/jmkelly/miniconda3/envs/moose
  registry file: /Users/jmkelly/.conda/environments.txt
done
(base) JMKs-MBP:~ jmkelly$ conda create -n moose moose-libmesh moose-tools
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: / failed

CondaError: KeyboardInterrupt

Again, I had to kill it.

**Reply by cticenhour** [reply]
Could you perform a `conda env list` and post the output? I'm wondering if it fully deleted your MOOSE environment.

**Reply by joe61vette** [reply]
Here it is:

(base) JMKs-MBP:~ jmkelly$ conda env list
# conda environments:
#
base                  *  /Users/jmkelly/miniconda3

**Reply by cticenhour** [reply]
OK so it looks like it was removed properly. One more thing to check `ls ~/miniconda3/envs`

If you don't see `moose` there, then I would try `conda create -n moose moose-libmesh moose-tools` again. It can take some time (several minutes or more) to do that "Solving Environment" step, especially if it has to do some extra work (since it failed the first time it tried).

**Reply by joe61vette** [reply]
I just tried to update Conda (just to make sure that wasn't contributing to the problem).  I got the same warning about being unable to register the environment.  Here it is:

I used:   conda update -n base -c defaults conda

and the warning was:

Downloading and Extracting Packages
ca-certificates-2021 | 121 KB    | ######################################################################################## | 100% 
libedit-3.1.20191231 | 96 KB     | ######################################################################################## | 100% 
certifi-2020.12.5    | 140 KB    | ######################################################################################## | 100% 
urllib3-1.26.3       | 105 KB    | ######################################################################################## | 100% 
openssl-1.1.1j       | 2.2 MB    | ######################################################################################## | 100% 
pysocks-1.7.1        | 31 KB     | ######################################################################################## | 100% 
sqlite-3.33.0        | 1.3 MB    | ######################################################################################## | 100% 
conda-4.9.2          | 2.9 MB    | ######################################################################################## | 100% 
idna-2.10            | 52 KB     | ######################################################################################## | 100% 
cffi-1.14.5          | 214 KB    | ######################################################################################## | 100% 
libffi-3.3           | 44 KB     | ######################################################################################## | 100% 
ruamel_yaml-0.15.87  | 235 KB    | ######################################################################################## | 100% 
brotlipy-0.7.0       | 333 KB    | ######################################################################################## | 100% 
cryptography-3.3.1   | 559 KB    | ######################################################################################## | 100% 
pycosat-0.6.3        | 86 KB     | ######################################################################################## | 100% 
pycparser-2.20       | 94 KB     | ######################################################################################## | 100% 
tqdm-4.56.0          | 80 KB     | ######################################################################################## | 100% 
yaml-0.2.5           | 72 KB     | ######################################################################################## | 100% 
chardet-4.0.0        | 194 KB    | ######################################################################################## | 100% 
conda-package-handli | 1.5 MB    | ######################################################################################## | 100% 
requests-2.25.1      | 52 KB     | ######################################################################################## | 100% 
tk-8.6.10            | 3.0 MB    | ######################################################################################## | 100% 
setuptools-52.0.0    | 721 KB    | ######################################################################################## | 100% 
pyopenssl-20.0.1     | 49 KB     | ######################################################################################## | 100% 
six-1.15.0           | 27 KB     | ######################################################################################## | 100% 
readline-8.1         | 333 KB    | ######################################################################################## | 100% 
Preparing transaction: done
Verifying transaction: / WARNING conda.core.path_actions:verify(963): Unable to create environments file. Path not writable.
  environment location: /Users/jmkelly/.conda/environments.txt

done
Executing transaction: \ WARNING conda.core.envs_manager:register_env(52): Unable to register environment. Path not writable or missing.
  environment location: /Users/jmkelly/miniconda3
  registry file: /Users/jmkelly/.conda/environments.txt
done

**Reply by cticenhour** [reply]
Looks like there might be a permissions issue for the `.conda` directory. Run `ls -la ~` and post the line relating to the `miniconda3` and `.conda` directories. For example, on my machine:

```
drwxr-xr-x    3 cticenhour  staff           96 Jul 20  2020 .conda
drwxr-xr-x   15 cticenhour  staff          480 Nov 16 16:33 miniconda3
```

The first bit of each line relates to the folder permissions. If this is not the case on your machine, then we'd need to try changing the permissions back to something sane. [This post's top answer on StackOverflow](https://stackoverflow.com/questions/59619442/how-does-one-fix-the-issue-of-not-writable-paths-with-conda) suggests performing the command `chmod 775 ./.conda` in your home directory.

**Reply by joe61vette** [reply]
First, let me say thanks for all the help Casey.  Here you go:

drwxr-xr-x   3 root     staff        96 Apr 25  2019 .conda
drwxr-xr-x  15 jmkelly  staff       480 Feb 24 17:56 miniconda3

This was installed in my admin user that has root.  They look the same as yours except for the "root".

**Reply by joe61vette** [reply]
I looked at the moose Conda trouble shooting page.  I did notice that the channels have changed since I installed this.  I can fix that once we get this figured out.

**Reply by cticenhour** [reply]
So the `root` owns that folder, which explains the errors. If you used `sudo` accidentally for an update or environment creation command, that might explain why, but then doesn't explain why `miniconda3` is owned by you and not `root`. Regardless, we can set the ownership of the folder to you, and hopefully fix this issue. 

Perform the command `sudo chown -R jmkelly ~/.conda`. Normally we wouldn't need to use `sudo` in your home directory (and shouldn't in general), but we need to do so in order to change the ownership of that directory.

**Reply by cticenhour** [reply]
Then please give me the same output from before showing the permissions. If that is now in your name, then you can try performing the `conda create` command again.

**Reply by joe61vette** [reply]
That worked!  It now shows:

drwxr-xr-x   3 jmkelly  staff        96 Apr 25  2019 .conda

Should I go back and run the update to Conda again?

**Reply by cticenhour** [reply]
Yes! Go for it - crossing my fingers this time that it works 😄

**Reply by joe61vette** [reply]
The Conda update worked (I think).  It showed:

conda update -n base -c defaults conda
Collecting package metadata (current_repodata.json): done
Solving environment: done

# All requested packages already installed.

But the conda create -n moose moose-libmesh moose-tools is still running.  So far:

(base) JMKs-MBP:~ jmkelly$ conda create -n moose moose-libmesh moose-tools
Collecting package metadata (current_repodata.json): done
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: / 

Still spinning....

**Reply by joe61vette** [reply]
Still spinning.  When I checked the channels, I got:

(base) JMKs-MBP:~ jmkelly$ conda config --show channels
channels:
  - https://mooseframework.org/conda/moose
  - conda-forge
  - defaults

Do I need to fix this before issuing the Conda create command?

**Reply by milljm** [reply]
If you wish to cancel that (taking to long), I have found installing one package at a time helps. Its the moose-tools package which I believe is holding it up (the moose-tools package has many dependencies conda has to solve). Try this instead:

```bash
conda create -n moose moose-libmesh
conda activate moose
conda install moose-tools
```

**Reply by cticenhour** [reply]
Since the conda update finished successfully, let's make sure your channels are right before going any farther. Stop that process and run `conda config --show channels`. On my machine, this outputs:
```
channels:
  - idaholab
  - conda-forge
  - defaults
  ```

**Reply by milljm** [reply]
> Still spinning. When I checked the channels, I got:
> 
> (base) JMKs-MBP:~ jmkelly$ conda config --show channels
> channels:
> 
>     * https://mooseframework.org/conda/moose
> 
>     * conda-forge
> 
>     * defaults
> 
> 
> Do I need to fix this before issuing the Conda create command?

Ahh.. that is a very old channel. Remove that, and add the new one:

```bash
conda config --remove channels https://mooseframework.org/conda/moose
conda config --add channels idaholab
```

**Reply by cticenhour** [reply]
Ah, the page just updated so I could see your message. Let's change your channels! 

Run `conda config --remove channels https://mooseframework.org/conda/moose` and then 
`conda config --add channels idaholab`

Then try to create the environment again.

**Reply by joe61vette** [reply]
Thanks Casey & Jason.  Appears to be working now.  It is downloading moose-tools at the moment.  Once it finishes, I'll go back to my clone of moose and try to compile.  Ah, it just finished.  Here goes...

**Reply by joe61vette** [reply]
And it compiled!!  Thanks so much for the help!  Can you mark this as "answered"?

**Reply by cticenhour** [reply]
I can - I will edit to expand on my original answer to make sure folks know that in this case, it was also a permissions issue.

**Reply by joe61vette** [reply]
Great.  Thanks again Casey!
